### PR TITLE
Get `console` messages to Go code

### PR DESCRIPTION
**Pull request by stroiman**
This adds the ability to get `console` messages to Go, but to do that you need to tap into the "inspector", which is also AFAIK the entry point to controlling the debugger. I just need the console messages. Although it seems that I can also use it to control time; which would be useful for my purpose.

This is still WIP, functionally close to done (for my needs), but I'd appreciate feedback, and opinion on code structure.

I also intend to add code-doc before merging.

**Inheriting from InspectorClient**

In C++, you must create a class inheriting from `InspectorClient`. This must then be installed using `Inspector::Create`, and you need to manually attach to each v8 context.

The `InspectorClient` has all virtual functions, and the client code overrides the ones they care about.

My idea was to create a Go interface for each capability. The user can choose to implement the functions they care about.

So far, I only care about console messages, so that's what I have now. It it still lacking stack traces, but should be simple enough.

When registering an inspector, you need a name, and a contextGroupId. I have no idea what they are used for, except contextGroupId must be <> 0, otherwise nothing happens. Right now they are hardcoded in C++ code.

That also means that I don't know if it makes sense to control from Go code or not.

The client itself does not seem to be related to an isolate. I don't know if it's valid to use the same client for multiple isolates. The code seems to indicate it's possible.

**Utility types**

The code uses a `StringView` type for strings. In reality; they're always utf16le (i.e. no byte order mark). Nothing is documented, so all is from experimenting, and only on one platform, my M2 mac. Btw, there must be a better way of converting `[]byte` to `[]uint16`, but I didn't find anything. Btw, I will add a test case that uses some weird unicode characters that take up more than one element in utf16. I did test that manually, and it did work, but better have it in the tests.

I created a more generic type for registering "callbacks". Just like the `Isolate` registers function callbacks, which I think should just be replaced with that type; which should be moved somewhere else.

**Package name**

If we follow the idea to create a separate interface for each capability that the inspector supports, it will be quite a lot of types, not necessarily with a sensible name of their own without some context. Prefixing them with `InspectorClient` makes them even longer, they're long enough IMHO.

I was wondering if it would make sense to put it in a sub-package, `v8go/inspector`. The C++ code does have it's own namespace, `v8_inspector`.

p.s. it's only now I see that I have reformatted `function_template.go`. I have just opened and saved it

### CODE REVIEW

**Comment by stroiman** [comment]
Btw, the initial code here predates the previous PR. I noticed I have created `InspectorPtr` and `InspectorClientPtr` types, which really isn't necessary, and I'm in favour of just replacing with `*v8Inspector` / `*v8InspectorClient`.

If you expressed your opinion on that topic, I have missed that.

**Comment by tommie** [comment]
> Btw, the initial code here predates the previous PR. I noticed I have created `InspectorPtr` and `InspectorClientPtr` types, which really isn't necessary, and I'm in favour of just replacing with `*v8Inspector` / `*v8InspectorClient`.
> 
> If you expressed your opinion on that topic, I have missed that.

No, I didn't. I don't see the point of the Ptr types, but I also don't have a strong opinion.

**Comment by stroiman** [comment]
Allright, I'll replace it with normal pointers, IMHO it's a code convention that adds more code in type declarations with no apparent benefit. I can easily imagine it's a result of finally finding a solution that compiles after struggling with the Go-compatible C-types vs. C++ types problem.

**Comment by stroiman** [comment]
Finally got around to update this.

I added documentation to types, replaced the `registry` with `cgo.Handle` types, and removed the `...Ptr` types.

While I'm in this branch, I'll just check to see if I can control time.

https://v8.github.io/api/head/classv8__inspector_1_1V8InspectorClient.html#ac27be4891cda5413f6156df6c98dc831

I assume that this controls the time reported by `new Date()` in JS land, which would be very useful to control from my purpose.

While that function use a `double` to represent time, I would use a Go [`time.Time`](https://pkg.go.dev/time#Time) type, as that would be the idiomatic way to represent a point in time.

**Review by tommie — COMMENTED** [review]
Please add a CHANGELOG entry for this added functionality.

File: inspector.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+func stringViewToString(d C.StringViewData) string {
+	if d.is8bit {
+		data := C.GoBytes(d.data, d.length)
+		return string(data)
+	} else {
+		data := C.GoBytes(d.data, d.length*2)
+		shorts := make([]uint16, len(data)/2)
+		for i := 0; i < len(data); i += 2 {
+			shorts[i/2] = (uint16(data[i+1]) << 8) | uint16(data[i])
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.go:117]: I don't know. It's not documented. https://v8.github.io/api/head/classv8__inspector_1_1StringView.html

I mentioned this in the original message, that in practice I only ever saw utf16le (without BOM) but that's long ago now. But looking at that again, I notice I mentioned an intent to add a test case for characters taking up more than 2 bytes for a single character.

I'll be sure to add that too.

File: inspector_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	defer iso.Dispose()
+	context := iso.NewContext()
+	defer context.Dispose()
+
+	_, err := context.RunScript("console.log('Hello, world!'); console.error('Error, world!');", "")
+	if err != nil {
+		t.Error("Error occurred: " + err.Error())
+		return
+	}
+	if len(recorder.messages) != 2 {
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector_test.go:69]: Much clearer :)

I've been a long time fan of gomega assertion library, so I'm not so used to base go test patterns.

File: inspector_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	t.Parallel()
+	recorder := consoleAPIMessageRecorder{}
+	iso := NewIsolateWithInspectorClient(&recorder)
+	defer iso.Dispose()
+	context := iso.NewContext()
+	defer context.Dispose()
+
+	_, err := context.RunScript("console.log('Hello, world!'); console.error('Error, world!');", "")
+	if err != nil {
+		t.Error("Error occurred: " + err.Error())
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector_test.go:66]: Good point

File: inspector.cc [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+using namespace v8;
+using namespace v8_inspector;
+
+class InspectorClient : public V8InspectorClient {
+  int _callbackRef;
+
+ public:
+  InspectorClient(int callbackRef) { _callbackRef = callbackRef; }
+  void consoleAPIMessage(int contextGroupId,
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.cc:15]: You mean the `consoleAPIMessage` function? I override a function in the base class, so it needs to have that name.

File: inspector.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
+package v8go
+
+// #include "inspector.h"
+import "C"
+import (
+	"sync"
+	"unicode/utf16"
+)
+
+type MessageErrorLevel uint8
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.go:14]: Good point. For the values, I adapted how e.g., `PropertyAttribute` is also implemented.

I think even linking to v8 documentation would make sense, but this doesn't look like a stable URL>

https://v8.github.io/api/head/classv8_1_1Isolate.html#acfc7c4d5c93913201249045e2655d3dd

File: inspector_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector_test.go]: So much clearer that I added all log levels to the test. Also added a basic string formatting for the levels.

File: inspector.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.go:137]: It _could_ be platform dependent. Checking the wiki page for v8, it supports more architectures than I was aware of, including PowerPC, which I believe is a big-endian architecture.

But is there a way to deal with this reliably?

https://en.wikipedia.org/wiki/V8_(JavaScript_engine)

File: inspector.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+typedef struct v8Isolate v8Isolate;
+
+typedef _Bool bool;
+
+#endif
+
+#include <stddef.h>
+#include <stdint.h>
+
+typedef struct m_ctx m_ctx;
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.h:33]: I wouldn't do that. The header file just needs to know a struct with that name _exists_, in order to define a pointer to it. The implementation need to know the type.

But I saw I had the `ContextPtr` type in here, which I removed.

File: inspector.cc [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+/********** InspectorClient **********/
+
+InspectorClientPtr NewInspectorClient(int callbackRef) {
+  return new InspectorClient(callbackRef);
+}
+
+void InspectorContextCreated(InspectorPtr inspector, ContextPtr context) {
+  LOCAL_CONTEXT(context);
+  int groupId = 1;
+  StringView name = StringView((const uint8_t*)"Test", 4);
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.cc:67]: Hmmm. It's called "humanReadableName", though I haven't found out where it's used. But I didn't realise you can create a `StringView` without constructor args.

The `groupId` appears to be sent back to the caller in the callback, `consoleAPIMessage`. It must be non-zero (no messages are received if it's zero).

Maybe I should expose both of these as arguments to the caller, but I was preferring a more minimal interface, rather than forcing the caller to pass arguments it has no use for.

File: inspector.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tommie) [reply inspector.go:136]: Looking around, it seems the 16-bit version is used for "protocol strings" and is cast to https://github.com/v8/v8/blob/main/src/inspector/string-16.h without conversion.

Looking in that implementation file, there's

https://github.com/v8/v8/blob/42da1a9ae8b7fa045232743dc810a5031f4a07ea/src/inspector/string-16.cc#L236-L249

which suggests that these are always stored LE.

I have to wonder how many String implementations a single project needs.

File: inspector.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (tommie) [reply inspector.h:33]: I based that suggestion on https://google.github.io/styleguide/cppguide.html#Forward_Declarations, but it's not a strong opinion.

**Review by tommie — COMMENTED** [review]
Sorry for the delay.

**Comment by stroiman** [comment]
> Sorry for the delay.

He, me too, I think there are still a few things I wanted to follow up on, in particular having a test case with special characters that have larger utf16-sequences.

**Comment by stroiman** [comment]
For the utf-16->utf-8 decoding, I had two ideas

- Would it make sense to do the decoding in C++ instead of Go? But I have no idea how to do that.
- The C++ code could prefix the string 0xFEFF (check it's not already there). That way, Go code just needs to treat it as UTF-16, and can forget about endianness.

However, I'm not sure that is the right way. Looking at `v8-inspector.h`, I see this

```
  // TODO(dgozman): add DCHECK(m_is8Bit) to accessors once platform can be used
  // here.
  const uint8_t* characters8() const { return m_characters8; }
  const uint16_t* characters16() const { return m_characters16; }
```

This indicates that the `Platform` is supposed to control this (eventually). So it would appear that we _should_ be able to control the encoding, and therefore be able to make assumptions about it too.

File: inspector.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.go:17]: Done!

File: inspector.go
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.go:1]: I assume, you mean the "exported to C" code function? `goHandleConsoleAPIMessageCallback`? Done!

File: inspector.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+type ConsoleAPIMessage struct {
+	contextGroupId int
+	ErrorLevel     MessageErrorLevel
+	Message        string
+	Url            string
+	LineNumber     uint
+	ColumnNumber   uint
+	// stackTrace StackTrace
```
[the remark this answers is not in the document]
Comment (stroiman) [reply inspector.go]: I added a comment about how the fields map to C++ function arguments, and a link to the `InspectorClient` C++ class documentation.

**Comment by stroiman** [comment]
Regarding the context group id specified here (where I found that it doesn't work with zero-value)

```
void InspectorContextCreated(v8Inspector* inspector, ContextPtr context) {
  LOCAL_CONTEXT(context);
  int groupId = 1;
  V8ContextInfo info = V8ContextInfo(local_ctx, groupId, StringView());
  inspector->contextCreated(info);
}
```

There are functions on the `Inspector` that use the group id, e.g. connecting an inspector session (which seems to just be a different name for a debugger). And `resetContextGroup`, whatever that does.

My conclusion is that the intention is that e.g., for a web page with `<iframe>`s, each frame has a separate context, but participate in the same context group, so one session can monitor the entire page.

This is just FYI, just sharing what I learned about the apparent purpose of these undocumented values are used. Didn't yet find where the "human readable name" is used, but didn't specifically look for it either.

File: inspector.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (tommie) [reply inspector.cc]: Thanks for investigating that (in another comment.) I just thought "Test" was an odd name for something in production. :)

Happy to leave it empty for now.

**Review by tommie — APPROVED** [review]

**Comment by tommie** [comment]
Again, thank you for taking the time!

Unfortunately, #81 is now blocking libv8 builds so the automatic release won't happen.

**Comment by stroiman** [comment]
np :)

I owe it to the project that my enhancements are made available.

I have a new batch of changes ready for a PR - it's not _all_ of my enhancements that are good for merge, some still need tests (they are implicitly tested from my other project).

They affect `FunctionTemplate`, `ObjectTemplate`, and `Object`, so part of the changes are also splitting up the .cc and .h files.

I've made a clean git history, so the first 3 commits are splitting up of files, and then it's one feature pr. commit.

Would you prefer one PR, or several?
